### PR TITLE
Job delete wait

### PR DESCRIPTION
**Pull request by ajayanto**
https://github.com/oneconvergence/gpuaas/issues/6506
https://github.com/oneconvergence/gpuaas/pull/6564
https://github.com/oneconvergence/gpuaas/pull/6739

### CODE REVIEW

**Comment by ajayanto** [comment]
@songole @mak-454  QA reported some issues with this fix. I have to check that. please don't merge it for now

**Comment by ajayanto** [comment]
Verified with QA it is working

File: dkube/sdk/api.py [hunk 1]
```diff
@@ -202,7 +202,25 @@ def list_ides(self, user, shared=False, filters='*'):
 
         return super().list_ides('notebook', user, shared)
 
-    def delete_ide(self, user, name):
+    def wait_for_delete_completion(self, uuid, jobclass, name):
```
Comment (mak-454) [inline dkube/sdk/api.py:205]: can you put this function at last and since this is internal function use "_wait_for_delete_completion"

File: dkube/sdk/api.py [hunk 2]
```diff
@@ -202,7 +202,25 @@ def list_ides(self, user, shared=False, filters='*'):
...
+        while True:
+            try:
+                data = super().job_get_by_uuid(uuid)
+                state = data['parameters']['generated']['status']['sub_state']
+                if state.lower() in ['deleting']:
```
Comment (mak-454) [inline dkube/sdk/api.py:210]: change to "=="

File: dkube/sdk/api.py [hunk 3]
```diff
@@ -283,7 +305,7 @@ def create_training_run(self, run: DkubeTraining, wait_for_completion=True):
                     "run {} - waiting for completion, current state {}".format(run.name, state))
                 time.sleep(self.wait_interval)
 
-    def get_training_run(self, user, name):
+    def get_training_run(self, user, name, fields='*'):
```
Comment (mak-454) [inline dkube/sdk/api.py:308]: why this change ?

File: dkube/sdk/api.py [hunk 4]
```diff
@@ -300,7 +322,7 @@ def get_training_run(self, user, name):
 
         """
 
-        return super().get_run('training', user, name)
+        return super().get_run('training', user, name, fields)
```
Comment (mak-454) [inline dkube/sdk/api.py:325]: this one

File: dkube/sdk/internal/api_base.py [hunk 5]
```diff
@@ -367,3 +367,7 @@ def download_model(self, path, user, name, version):
         copy_tree("{}/{}/data".format(tempdir.name, version), path)
         # use temp_dir, and when done:
         tempdir.cleanup()
+    def job_get_by_uuid(self, uuid):
```
Comment (mak-454) [inline dkube/sdk/internal/api_base.py:370]: did you add this manually?

**Comment by mak-454** [comment]
@ajayanto am not sure but is PR on latest 2.2 ?
also can you run **autopep8 -i <filename>**

File: dkube/sdk/api.py (hunk 3 above)
Comment (ajayanto) [reply dkube/sdk/api.py:308]: It is a change done by @songole  in his patch to get the entire response or the state

File: dkube/sdk/internal/api_base.py (hunk 5 above)
Comment (ajayanto) [reply dkube/sdk/internal/api_base.py:370]: The ._api.jobs_get_one_by_uuid was there. but the function in api_base was missing so I added it.

Comment (ajayanto) [reply dkube/sdk/internal/api_base.py:370]: It is just a wrapper function. Can it be generated without manually adding

File: dkube/sdk/api.py (hunk 1 above)
Comment (ajayanto) [reply dkube/sdk/api.py]: done

File: dkube/sdk/api.py (hunk 2 above)
Comment (ajayanto) [reply dkube/sdk/api.py]: done

**Comment by ajayanto** [comment]
Yes it is on the latest 2.2

File: dkube/sdk/api.py (hunk 3 above)
Comment (mak-454) [reply dkube/sdk/api.py:290]: is this required and relevant to this pr ?

File: dkube/sdk/api.py (hunk 4 above)
Comment (mak-454) [reply dkube/sdk/api.py:307]: is this required and relevant to this pr ?

**Comment by mak-454** [comment]
this PR cannot be merged until the BE release is made.

**Comment by mak-454** [comment]
@songole made the wait_for_completetion=False by default so that the SDK func works with older release (backward compatability).
From 2.2.7.x onwards the client code would need explicitly wait_for_completion=True.
@rahul-179 the automation code should pass this flag == True when testing from 2.2.7.x release onwards.

**Comment by songole** [comment]
If substate is not set to deleting or deleted, can we assume it is 2.1?

**Comment by mak-454** [comment]
thought of it as if substate is set to empty then assume old release, but i was worried that if theres any corner case - then retry covers it.

**Comment by mak-454** [comment]
@songole handled by checking 404 code. Old release(s) will return 404, for new release it should work.

**Comment by mak-454** [comment]
@rahul-179 no changes needed - ignore my previous comment.
